scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -649,294 +661,134 @@ def dijkstra(csgraph, directed=True, indices=None,
+        # the vertex v.second
+        # This assures _dijkstra_scan_heap is only called once per vertex
+        # and the total complexity is O(Mlog(M)) per source
+        if dist_matrix[v.second] < v.first :
+            continue
 
-            _dijkstra_scan_heap(&heap, v, nodes,
+        _dijkstra_scan_heap(heap, v, csr_weights, csr_indices, csr_indptr,
+                            dist_matrix, pred, return_pred,
+                            sources, return_sources, limit)
+        if not directed:
+            _dijkstra_scan_heap(heap, v,
                                 csrT_weights, csrT_indices, csrT_indptr,
-                                pred, return_pred, limit, i)
-
-            # v has now been scanned: add the distance to the results
-            dist_matrix[i, v.index] = v.val
-
-    free(nodes)
+                                dist_matrix, pred, return_pred,
+                                sources, return_sources, limit)
     return 0
 
 @cython.boundscheck(False)
-cdef int _dijkstra_undirected_multi(
+cdef int _dijkstra_multi_separate(
             const int[:] source_indices,
             const double[:] csr_weights,
             const int[:] csr_indices,
             const int[:] csr_indptr,
             const double[:] csrT_weights,
             const int[:] csrT_indices,
             const int[:] csrT_indptr,
-            double[:] dist_matrix,
-            int[:] pred,
-            int[:] sources,
+            double[:, :] dist_matrix,
+            int[:, :] pred,
+            int[:, :] sources,
             DTYPE_t limit) except -1:
     cdef:
-        unsigned int N = dist_matrix.shape[0]
-        int return_pred = (pred.size > 0)
-        FibonacciHeap heap
-        FibonacciNode *v
-        FibonacciNode* nodes = <FibonacciNode*> malloc(N *
-                                                       sizeof(FibonacciNode))
-    if nodes == NULL:
-        raise MemoryError("Failed to allocate memory in "
-                          "_dijkstra_undirected_multi")
-
-    _dijkstra_setup_heap_multi(&heap, nodes, source_indices,
-                               sources, dist_matrix, return_pred)
-
-    while heap.min_node:
-        v = remove_min(&heap)
-        v.state = SCANNED
-
-        _dijkstra_scan_heap_multi(&heap, v, nodes,
-                                  csr_weights, csr_indices, csr_indptr,
-                                  pred, sources, return_pred, limit)
-
-        _dijkstra_scan_heap_multi(&heap, v, nodes,
-                                  csrT_weights, csrT_indices, csrT_indptr,
-                                  pred, sources, return_pred, limit)
-
-        #v has now been scanned: add the distance to the results
-        dist_matrix[v.index] = v.val
+        unsigned int Nind = source_indices.shape[0]
+        unsigned int i
+        int source_list[1]
 
-    free(nodes)
-    return 0
+    assert dist_matrix.shape[0] == Nind
+    assert pred.shape[0] == Nind
+    assert sources.shape[0] == Nind
```

Comment:
Again, raw asserts in code, please change to checks and throw an exception if this shouldn't occur